rust/src/dataset/scanner.rs

```diff
@@ -40,79 +52,118 @@ pub struct Scanner<'a> {
     offset: Option<i64>,
 
     fragments: Vec<Fragment>,
-
-    // use for iterator
-    fragment_idx: usize,
-    batch_id: i32,
-    reader: Option<FileReader<'a>>,
 }
 
 impl<'a> Scanner<'a> {
-    pub fn new(dataset: &'a Dataset) -> Result<Self> {
-        Ok(Self {
+    pub fn new(dataset: &'a Dataset) -> Self {
+        Self {
             dataset,
             projections: dataset.schema().clone(),
             limit: None,
             offset: None,
             fragments: dataset.fragments().to_vec(),
-
-            fragment_idx: 0,
-            batch_id: 0,
-            reader: None,
-        })
+        }
     }
 
+    /// Projection.
+    ///
+    /// Only seelect the specific columns. If not specifid, all columns will be scanned.
     pub fn project(&mut self, columns: &[&str]) -> Result<&mut Self> {
         self.projections = self.dataset.schema().project(columns)?;
         Ok(self)
     }
 
+    /// Set limit and offset.
     pub fn limit(&mut self, limit: i64, offset: Option<i64>) -> &mut Self {
         self.limit = Some(limit);
         self.offset = offset;
         self
     }
 
-    pub async fn next_batch(&mut self) -> Option<Result<RecordBatch>> {
-        if self.fragment_idx >= self.fragments.len() {
-            return None;
-        }
+    /// The schema of the output, a.k.a, projection schema.
+    pub fn schema(&self) -> SchemaRef {
+        Arc::new(ArrowSchema::from(&self.projections))
+    }
 
-        let fragment = &self.fragments[self.fragment_idx];
-        // Only support 1 data file (no schema evolution for now);
-        assert!(fragment.files.len() == 1);
-        let data_file = &fragment.files[0];
-        let path = self.dataset.data_dir().child(data_file.path.clone());
-        if self.reader.is_none() {
-            self.reader = Some(
-                FileReader::new(
-                    &self.dataset.object_store,
-                    &path,
-                    Some(&self.dataset.manifest),
-                )
-                .await
-                .unwrap(),
-            );
-            self.reader
-                .as_mut()
-                .map(|reader| reader.set_projection(self.projections.clone()));
-        }
+    /// Create a stream of this Scanner.
+    ///
+    /// TODO: implement as IntoStream/IntoIterator.
+    pub fn into_stream(&self) -> ScannerStream {
+        const PREFECTH_SIZE: usize = 8;
+        let object_store = self.dataset.object_store.clone();
```

Comment:
this is cloning the Arc? Does it clone the underlying object store?

Reply:
`Arc::clone()` is like copying the C++ `shared_ptr`, it only copys the pointer  and increase the reference count, does not copy the actual data.